catalog/permissions/permissions.go

```diff
@@ -0,0 +1,152 @@
+package permissions
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/service/catalog"
+	"github.com/databricks/databricks-sdk-go/service/sharing"
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// API
+type UnityCatalogPermissionsAPI struct {
+	client  *databricks.WorkspaceClient
+	context context.Context
+}
+
+func NewUnityCatalogPermissionsAPI(ctx context.Context, m any) UnityCatalogPermissionsAPI {
+	client, _ := m.(*common.DatabricksClient).WorkspaceClient()
+	return UnityCatalogPermissionsAPI{client, ctx}
+}
+
+func (a UnityCatalogPermissionsAPI) GetPermissions(securable catalog.SecurableType, name string) (list *catalog.PermissionsList, err error) {
+	if securable.String() == "share" {
+		list, err = a.client.Shares.SharePermissions(a.context, sharing.SharePermissionsRequest{name})
+		return
+	}
+	list, err = a.client.Grants.GetBySecurableTypeAndFullName(a.context, securable, name)
+	return
+}
+
+func (a UnityCatalogPermissionsAPI) UpdatePermissions(securable catalog.SecurableType, name string, diff []catalog.PermissionsChange) error {
+	if securable.String() == "share" {
+		return a.client.Shares.UpdatePermissions(a.context, sharing.UpdateSharePermissions{
+			Changes: diff,
+			Name:    name,
+		})
+	}
+	_, err := a.client.Grants.Update(a.context, catalog.UpdatePermissions{
+		Changes:       diff,
+		SecurableType: securable,
+		FullName:      name,
+	})
+	return err
+}
+
+func (a UnityCatalogPermissionsAPI) WaitForUpdate(timeout time.Duration, securable catalog.SecurableType, name string, desired catalog.PermissionsList, diff func(*catalog.PermissionsList, catalog.PermissionsList) []catalog.PermissionsChange) error {
+	return retry.RetryContext(a.context, timeout, func() *retry.RetryError {
+		current, err := a.GetPermissions(securable, name)
+		if err != nil {
+			return retry.NonRetryableError(err)
+		}
+		log.Printf("[DEBUG] Permissions for %s-%s are: %v", securable.String(), name, current)
+		if diff(current, desired) == nil {
+			return nil
+		}
+		return retry.RetryableError(
+			fmt.Errorf("permissions for %s-%s are %v, but have to be %v", securable.String(), name, current, desired),
+		)
+	})
+}
+
+// Terraform Schema
+type UnityCatalogPrivilegeAssignment struct {
```

Comment:
Can you comment here that this is based off of `catalog.PermissionsList`? As @nkvuong mentioned, in the future we'll refactor these resources to inherit fields from the Go SDK. Knowing what this is derived from will help us track this down in the future.